src/PrestaShopBundle/Security/Admin/Employee.php

```diff
@@ -59,20 +55,15 @@ class Employee implements UserInterface, EquatableInterface
      */
     private $roles = array();
 
-    private $data;
-
     /**
-     * Constructor.
-     *
-     * @param object $data The employee legacy object
+     * @param AuthenticatedEmployee $authenticatedEmployee
      */
-    public function __construct($data)
+    public function __construct(AuthenticatedEmployee $authenticatedEmployee)
```

Comment:
Is this considered a BC break? `$data` argument was expecting a legacy `Employee` instance, even though it used only a few fields from it inside the class. With new authorization system I feel like this class needed this cleanup.

wdyt @matks?

Reply:
Yes it's a BC break 😬 
Moreover I think that we should not force people to have a `AuthenticatedEmployee` to be able to build the logged-in "PrestaShopBundle\Security\Admin\Employee".

What if you want to create a SSO for your shop, so authentication does not come from the database but from an API instead ?

So this array of `$data` is not bad actually 🤔 it is agnostic. It does not care where the data comes from.

Reply:
what I don't like is that `$data` is used as an employee object extensively later on, but injected without any typehint :confused: 
https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Security/Voter/PageVoter.php#L77
https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Security/Admin/Employee.php#L71

Reply:
indeed that is nasty :/
but we should aim for agnostic data

`PrestaShop\PrestaShop\Core\Domain\Employee\QueryResult\AuthenticatedEmployee` is too coupled to Domain/CQRS/Core. Tomorrow, an agency developer should be able to create his own authentication system (like facebook login, or twitter login) without having to use this class.
Legacy `Employee` class is not good either. Same issue.

What we need to do is to define **what an authenticated user in PrestaShop BO should have as properties**. We can fill these data however we want. Maybe it'll look 90% like legacy Employee. But we need to make it independent.

Reply:
Good point - I agree with you. However, can I change the constructor signature though, to inject only the needed properties, or does it have to stay as `$data`?

Reply:
No 😬 BC break
We have to live with the mistakes of the past

What you can do however is add a static builder function
```
public static function createFromFooBar(FooBar $fooBar)
{
     return new self(
        $fooBar->getX(),
        $fooBar->getY(),
        $fooBar->getZ(),
}
```